common/lib/new.cc

```diff
@@ -0,0 +1,369 @@
+            last_block_node->m_next->m_prev = last_block_node;
+          }
+          m_used -= sizeof(block_node);
+          m_free += sizeof(block_node);
+          continue;
+        }
+      }
+
+      last_block = &block;
+    }
+  }
+
+  void set_new_limit(std::size_t new_limit) {
+    if (new_limit >= (m_heap_limit + MIN_ALLOC_SIZE)) {
+      add_block(
+          // new block address will be directly after the region defined by the
+          // previous limit.
+          &__heap_start + m_heap_limit, new_limit - m_heap_limit);
+
+      m_heap_limit = new_limit;
+    }
+  }
+
+private:
+  // nodes are added at the end of the list.
+  void add_node(block_node *node) {
+    list_node *node_last = m_head;
+    for (; node_last->m_next; node_last = node_last->m_next) {
+    }
+
+    node_last->m_next = node;
+    node->m_prev = node_last;
+  }
+
+  void add_block(std::byte *address, std::size_t size) {
+    // assume zero alignment padding.
+    add_node(new (address) block_node{size});
+    m_used += sizeof(block_node);
+    m_free += m_head->m_block.m_size;
+  }
+
+  block_node *m_head = nullptr;
+
+public:
+  // heap limit is essentially a global var.  It can
+  // be set prior to ctor of blocklist.
+  static constexpr std::size_t HEAP_DEFAULT_LIMIT = 4096;
+  static std::size_t m_heap_limit;
+
+  // accounting.
+  std::size_t m_free = 0;
+  std::size_t m_used = 0;
+};
+
+blocklist &get_free_list() noexcept {
+  static blocklist free_list;
+  return free_list;
+}
+
+void run_new_handler() {
+  const auto newp = std::get_new_handler();
+  if (newp) {
+    newp();
+  }
+}
+} // namespace
+
+std::size_t blocklist::m_heap_limit = SIZE_MAX;
+
+void *operator new(std::size_t size) {
+  const auto block = operator new(size, std::nothrow);
+  if (!block) {
+    run_new_handler();
+  }
+
+  return block;
+}
+
+void *operator new[](std::size_t size) { return operator new(size); }
+
+void *operator new(std::size_t count, const std::nothrow_t &) noexcept {
+  auto &free_list = get_free_list();
+  return free_list.split_block(free_list.find_first_fit(count), count);
+}
+
+void *operator new[](std::size_t count, const std::nothrow_t &) noexcept {
+  return operator new(count, std::nothrow);
+}
+
+void *operator new(std::size_t, void *ptr) { return ptr; }
+
+void *operator new[](std::size_t, void *ptr) { return ptr; }
+
+void operator delete(void *ptr) noexcept {
+  if (!ptr) {
+    return;
+  }
+
+  auto &free_list = get_free_list();
+  free_list.free_block(block::get_block(static_cast<std::byte *>(ptr)));
+  free_list.coallesce_blocks();
+}
+
+void operator delete[](void *ptr) noexcept { operator delete(ptr); }
+
+// helper functions for getting/setting flags in guard_object
+static bool initializerHasRun(std::uint8_t *guard_object) {
+  return (*guard_object != 0);
+}
+
+static void setInitializerHasRun(std::uint8_t *guard_object) {
+  *(guard_object) = 1;
+}
+
+extern "C" {
+int __cxa_guard_acquire(std::uint8_t *guard_object) {
```

Comment:
This doesn't actually provide the guarantees that __cxa_guard_acquire is supposed to in the presence of asynchronicity.

Example:
1. The main program enters a static initializer, calling __cxa_guard_acquire. initializerHasRun completes, and returns zero.
2. Before this can return, an NMI interrupt fires. The NMI handler calls the same function. Since nothing has yet been written back, initializerHasRun also returns 0. This causes the whole initialization to run on the NMI thread.
3. The NMI returns. The original __cxa_guard_acquire returns 1, and the whole initialization runs a second time, this time on the main thread.

Disabling interrupts doesn't really fix this, since NMI can call C code. It's possible to disable NMIs on cerrtain platforms (maybe all?), but the means for doing so is platform specific.

There's also the possiblity that this would be runnint in a more hosted environment, say, one that provides real threads and async signals. In that case, a real lock implementation might be on offer, and that would be usable here.

Accordingly, we could make a platform-specific set of lock or atomic primitives and use those to implement these. There's a couple ways I could see doing that:
1. Disable all interrupts using platform-specific mechanisms. If we came up against a plaform where there was no way to disable NMI, we'd be SOL if this specific scenario occurs. Even though it's rare, it would create pointy bits in the user experience that we've mostly been able to avoid so far.
2. Find a clever way to use the uninterruptable INC and DEC operations to implement a kind of semaphore or lock. If there's a way to do this, we could have a truly generic lock/atomic implementation. I'd prefer this if possible, since it keeps the complexity away from the user facing side of things, and it generalizes across platforms.
3. Implement one of the fancy algorithms (https://cs.stackexchange.com/questions/30247/are-there-hardware-lock-implementations-without-test-and-set-or-swap) for locking over basic primitives. I haven't looked into any of these in detail, but if they work as they claim on the tin, then it should be a workable fallback. Not sure if any of them are *fast* though.

Reply:
It occured to me after writing this that the situation is trickier for interrupt handlers than for traditional threads, since the interrupted control flow cannot resume until the interrupt returns. This means that any traditional lock held by an interrupted thread would deadlock if the same lock was required by the interrupting thread. For the static allocation case, if the main thread began initializing a function static variable, and an NMI needed access to the same static, there'd be absolutely no safe way to proceed. The NMI needs access to the static before it can proceed and return, but there's no way for the main thread to finish initializing it. The NMI can't safely initialize it either, since it's already in the process of being initialized.

This is something that will need more careful thought; perhaps an inspection of the behaviors of pre-existing freestanding C++ environments might be illuminating.

Reply:
From https://sudonull.com/post/139020-C-support-for-avr-in-gcc, one way to deal with this is to completely disable interrupts for the duration of the initialization. That is, disable them in `__cxx_guard_acquire`, and reenable them in `__cxx_guard_release`. This would imply having a platform-specific interrupt disabling mechanism, though.

Reply:
Oh, a cheesy way to disable NMI on any platform is just to point the system interrupt vector to an RTS for the duration, then restore it back. It's still not 100% ideal; if a static initializer or other process were to manipulate the NMI vector, then it would reenable interrupts, causing bad things. So we'd have to call out manipulating the NMI vector while a C++ static initialization is active as undefined behavior, which we're in our rights to do, since NMI manipulation can only be done via implementation-defined behavior.

Reply:
Ah, but changing the NMI vector is also usually platform specific. Que sera.

Reply:
Alllllternatively we could declare doing this specific thing in an NMI to be undefined behavior, just set the interrupt disable flag, and call it a day.

Reply:
Right; I'll have to think on this. Will get back to you on this later.

Reply:
OK-- I think you're over-thinking this.  This problem is not unique to freestanding implementation.  Hosted implementations also have interrupt service functions, for example consider signal handlers.  

The one-time construction API only provides *thread* safety, not *interrupt* safety.  If there is only a single thread of execution, we can reason that an interrupt handler cannot execute concurrently with the code that it interrupted. Therefore, any attempt for __cxa_guard_acquire to acquire the lock within the interrupt routine would fail/deadlock if the interrupt is triggered while the thread had already acquired the lock.

I am wagering that any implementation of __cxa_guard_acquire fails in the context of an interrupt service routine. I'm curious what happens if you try this on a regular, hosted clang++ implementation.  I'll try it.

I'm pretty sure disabling interrupts within the __cxa_guard_acquire is a non-starter, for any hosted or freestanding implementation. The one time initialization guard has to be invoked every time execution references the static variable so it has to be extremely cheap to test if the initialization has already occurred. Enabling/disabling interrupts every time you reference a static function scope variable would be both too expensive and too surprising. 

So; for this single-threaded implementation, the trivial non-implementation of __cxa_guard is probably the correct one.

Reply:
So, the standard was a bit tricky to interpret here; the language used doesn't say anything about thread safety; it just says:

> If control enters the declaration concurrently while the variable is being initialized, the concurrent execution shall wait for completion of the initialization.


Which would, prima facie, include any means by which concurrent execution could occur.

However, independently of this, a function is not signal-safe if, amongst other things, it is "waiting for the completion of the initialization of a variable with static storage". Performing a non-signal-safe evaluation in a signal handler (which we can generalize to interrupt handlers) is undefined behavior. So the wording of the above becomes moot.

Accordingly, we can go ahead and resolve this. Please do add a comment above these functions summarizing the discussion; this should help casual readers understand why we don't need a more sophisticated __cxa_guard.